### ISSUE TITLE
Gallery buttons cannot display icons normally

### ISSUE THREAD

**Issue by liraqb** [qsniyg/maxurl#347]
`Gallery buttons` on the left and right sides cannot display icons normally. The box symbol means that the icon is not loaded normally, right? 

![image](https://user-images.githubusercontent.com/62490653/82913974-8cdbfd80-9fa1-11ea-9445-6e151818e1a8.png)

**Comment by qsniyg** [comment]
Does this only happen on one website, or does it happen everywhere? The reason for this is because all of the icons are actually just emojis or unicode characters, so I'm guessing the font you're using doesn't support it.

Maybe there's a way to detect if the user's system supports a glyph, and if it doesn't, use an alternate one instead...

**Comment by liraqb** [comment]
This problem exists on every website.

I found symbols that cannot be displayed in Tampermonkey. I found that the download button also cannot be displayed normally.

Search keywords: `// \u`

![image](https://user-images.githubusercontent.com/62490653/82916946-4e484200-9fa5-11ea-8431-ca84c9165c50.png)

![image](https://user-images.githubusercontent.com/62490653/82916962-54d6b980-9fa5-11ea-8e81-9e3e5213dc84.png)

Modify `var icon = leftright ? "\ud83e\udc52" : "\ud83e\udc50";` to `var icon = leftright ? "\u2192" : "\u2190";`, `Gallery buttons` are displayed normally.

![image](https://user-images.githubusercontent.com/62490653/82920387-ca448900-9fa9-11ea-9591-aebf23f7a00b.png)

I don't know how to modify the symbol of the download button. I can't search for characters like `\ud83e\udc52`, what websites can search for it?

**Comment by qsniyg** [comment]
Could you check if the latest commit (https://github.com/qsniyg/maxurl/commit/5821c6184f1c8f124e3ba0b34fce7963690f2a39) fixes it?

I used this website: http://xahlee.info/comp/unicode_arrows.html to find the arrows :)

**Comment by liraqb** [comment]
It still doesn't show up properly.

**Comment by qsniyg** [comment]
Ah [I should have checked the spec more closely](https://lists.w3.org/Archives/Public/www-style/2015Aug/0330.html).

Since I can't find any way to fix this automatically, I've just added a new option named "use safe glyphs". Not a great solution, but better than nothing, hopefully :)

**Comment by liraqb** [comment]
Thanks, after enabling this option, the symbol is displayed normally.

## About Noto Sans font

I installed Noto Sans, restarted my PC, and changed all browser fonts to Noto Sans. The symbol still cannot be displayed normally. I use Windows 7 x64 Simplified Chinese operating system.

Noto Sans has been installed:

![image](https://user-images.githubusercontent.com/62490653/82979017-905e9b80-a018-11ea-9fc6-d87ba9aac21a.png)

Current browser fonts:

![image](https://user-images.githubusercontent.com/62490653/82979025-93598c00-a018-11ea-8ac3-dfd3d723c77d.png)

The symbols in the script code cannot be displayed normally:

![image](https://user-images.githubusercontent.com/62490653/82979029-96547c80-a018-11ea-8a1b-49f19797be81.png)

**Comment by qsniyg** [comment]
@liraqb Just curious, does it work with https://github.com/qsniyg/maxurl/commit/097780b69c09a416abab7455e7ca2011bb52fab8? Tampermonkey does have an issue with encoded unicode characters in strings, but it should _in theory_ work in the browser...

**Comment by liraqb** [comment]
The symbol still cannot be displayed. It also cannot display symbols on the options page.

![image](https://user-images.githubusercontent.com/62490653/83083225-2d244600-a0b8-11ea-9545-8d6452ad1ce9.png)

On the website http://xahlee.info/comp/unicode_arrows.html,  the font doesn't seem to work.

![image](https://user-images.githubusercontent.com/62490653/83083228-301f3680-a0b8-11ea-8218-47c9fd714a7f.png)

**Comment by Gitoffthelawn** [comment]
I gave this extension a try, and I have the same issue. Most of the glyphs don't display correctly in the extension's popup UI or settings page.

**Comment by Gitoffthelawn** [comment]
@qsniyg I gave a few things a try, and none worked. 

On a test system, I tried installing the *Noto Sans Regular* font from Google.  I then used `@font-face` to redefine `sans-serif` as *Noto Sans Regular*.  That didn't work.

I then forced the extension to use *Noto Sans Regular* by injecting this code:
```
html, body, div, span, applet, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
a, abbr, acronym, address, big, cite, code,
del, dfn, em, img, ins, kbd, q, s, samp,
small, strike, strong, sub, sup, tt, var,
b, u, i, center,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, embed,
figure, figcaption, footer, header, hgroup,
menu, nav, output, ruby, section, summary,
time, mark, audio, video, button {
	font-family: "Noto Sans Regular", sans-serif !important; }
```

That didn't work either, so I took a closer look at *Noto Sans Regular*.  I went onto https://fonts.google.com, and entered the glyph being used by *Image Max URL* for the *revert* symbol in the settings (as an example).  I scrolled through a few hundred of the Google fonts, and none of them (including all the *Noto Sans* fonts) appear to include that glyph (⮌) (\u2b8c).

Here is a link of a demonstration of that font trying to display that glyph:

https://fonts.google.com/specimen/Noto+Sans?preview.text=⮌&preview.text_type=custom&stylecount=1&query=noto

**Comment by qsniyg** [comment]
I had installed noto fonts from my package manager, and I had assumed it would include all fonts for everyone else, my bad! After looking at gucharmap (a program that lets you find which font is used for which character), turns out that it's in [Noto Sans Symbols2](https://www.google.com/get/noto/#sans-sym2).

Good idea about the svg thing though, I should definitely replace it with that. Been thinking of doing that for a while, but never really got around to it for some reason haha. I'll also add an option to allow it to be modified through the css styles.

**Comment by Gitoffthelawn** [comment]
Adding them as SVG paths sounds great! Thank you!

**Comment by Gitoffthelawn** [comment]
@qsniyg  I was thinking more about this, and there are several methods to accomplish this goal.  I think my 2 favorite methods are:

   1. Add `PATH` elements to draw the glyphs.  I'm thinking there are probably online converters to convert from fonts automatically.  If you find any good ones you like, please let me know.  Alternatively, you can simply use open-source glyphs defined as PATHs.
   2. Add `data:image/svg+xml` images to draw the glyphs.  I'm quite sure there are online converters to do this.

I think I prefer method 1 because it makes it very easy to see exactly what the code is doing and is also easy to edit.

Another advantage of both of these methods is that performance should be improved, as no additional font resources will be required (and the CSS can likely be simplified as well).

**Comment by qsniyg** [comment]
Personally I was actually about to go for option 2, just because it allows for very easy overriding through the button CSS styles option (e.g. `-imu-image: "data:image/svg+xml,..."`), allowing people to override the image to be whatever they want it to be. Of course I don't expect most users to do this, but it might be useful for themes and the likes (which is already possible by importing custom settings, with `+` at the beginning to only override the specified settings).

For seeing what the code is doing, you're right that an svg tag is easier, but I that data urls aren't necessarily too hard either. They're either url encoded (which can be decoded online or through `decodeURIComponent("...")` in the browser console), or base64 (again, either online or `btoa("...")`). Or you can also just paste the url in the browser, then save the result (or save it from the network tab in devtools). For editing, once it's saved, it should be just as easy because it's the same data. The advantage the data url has though, is that you can then use the css styles option in order to change the image for future popups.

If you disagree though, please let me know, these are just my thoughts on this. I think for the moment I'll go for the data urls just because they're also much easier to implement, but I have no problem adding an option or something if there is a use case for the other way too.

**Comment by Gitoffthelawn** [comment]
I agree with everything you wrote.  Both of the methods I proposed are my favorites for implementation.

Given your wonderful feedback, I really think it's a toss-up, and either way is excellent.

**Comment by qsniyg** [comment]
Ah, I just figured out a reason against 2 and for 1: No support for recoloring svgs (unless of course the data would be edited, which is a possibility, but is rather hacky and potentially somewhat performance-intensive). But this would be possible in the first option (svg tag), as the "fill" parameter could just be added. This is a bit of an issue because a few people have changed the background color of the buttons to be light instead of dark, and this would make it more or less impossible to use the buttons (since the icons would be whiteish).

Alternatively, it _might_ be possible to turn the image monochromatic and do some kind of post-processing filter on it. I don't know how good of an idea that is though.

I would still like to _support_ images to be clear, but I'm not sure if defaulting to them is the best idea. I'm still not sure what the best way to support svgs would be (maybe just through a `-imu-svg` property with the svg html? It's not ideal in a few ways, but I guess it could work). I'll leave this for a bit, see if a better solution comes to mind later. If you have any other ideas, please feel free to share!
